src/wrappers/themis/wasm/wasmthemis.mk

```diff
@@ -19,11 +19,13 @@ WASM_PATH = src/wrappers/themis/wasm
 WASM_SRC += $(WASM_PATH)/package.json
 WASM_SRC += $(wildcard $(WASM_PATH)/src/*.js)
 
+WASM_EXPORTED = $(abspath $(WASM_PATH)/emscripten/exported_functions.json)
 WASM_RUNTIME = $(abspath $(WASM_PATH)/emscripten/runtime_exports.json)
 WASM_PRE_JS  = $(abspath $(WASM_PATH)/emscripten/pre.js)
 
 WASM_PACKAGE = $(BIN_PATH)/wasm-themis.tgz
 
+$(BIN_PATH)/libthemis.js: LDFLAGS += -s EXPORTED_FUNCTIONS=@$(WASM_EXPORTED)
```

Comment:
Hmm... I’m pretty sure that using [EMSCRIPTEN_KEEPALIVE][1] with [THEMIS_API][2] macros should mark all public functions as exported and we don’t have to repeat them in EXPORTED_FUNCTIONS list. If that’s not working then probably [\_\_EMSCRIPTEN\_\_][3] macro is not defined which is weird.

I definitely remember having LLVM strip all our functions as dead code. EMSCRIPTEN_KEEPALIVE fixed that. And it seems to still work for me with the latest Emscripten 1.39.4 and 1.39.4-fastcomp.

Does webpack produce a warning or something if EMSCRIPTEN_KEEPALIVE is used instead of explicit EXPORTED_FUNCTIONS list?

[1]: https://emscripten.org/docs/api_reference/emscripten.h.html?highlight=emscripten_keepalive#c.EMSCRIPTEN_KEEPALIVE
[2]: https://github.com/cossacklabs/themis/blob/255acec5c815c0578523b2e215e91773cde75b33/src/themis/themis_api.h#L20-L36
[3]: https://emscripten.org/docs/compiling/Building-Projects.html#detecting-emscripten-in-preprocessor

Reply:
I am sorry. I am new to use Emscripten. I don't know you already have a solution to use EMSCRIPTEN_KEEPALIVE. Your solution is better if it work. I just git clone the code and follow the build steps. Everything seems OK without warning but the result cannot pass any test case. I think it is not related with webpack. Maybe I missed some requirements for Emscripten to enable those macros. I build it on macOS Mojave 10.14.6 and emcc version is as follows. 

emcc -v
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 1.38.44
clang version 10.0.0 (/b/s/w/ir/cache/git/chromium.googlesource.com-external-github.com-llvm-llvm--project b5f295ffcec2fa7402e39eb1262acbd55a7d39f5)
Target: x86_64-apple-darwin18.7.0
Thread model: posix

Reply:
I forgot to mention. There are two errors at the beginning of build. They are related with "PHP". I have no idea if the issue is the cause.
/bin/bash: line 1: 44577 Abort trap: 6           php -r "echo PHP_MAJOR_VERSION;" 2> /dev/null
/bin/bash: line 1: 44625 Abort trap: 6           php -r "echo PHP_MAJOR_VERSION.'.'.PHP_MINOR_VERSION;" 2> /dev/null

Reply:
I post all the output of build for the command "emmake make wasmthemis" without explicit EXPORTED_FUNCTIONS list. I check the result again and it has no exported functions and it cannot pass test cases.

cache:INFO: generating system asset: is_vanilla.txt... (this will be cached in "/Users/frank/.emscripten_cache/is_vanilla.txt" for subsequent builds)
cache:INFO:  - ok
make: ['make', 'wasmthemis']
/bin/bash: line 1: 44577 Abort trap: 6           php -r "echo PHP_MAJOR_VERSION;" 2> /dev/null
/bin/bash: line 1: 44625 Abort trap: 6           php -r "echo PHP_MAJOR_VERSION.'.'.PHP_MINOR_VERSION;" 2> /dev/null
compile build/obj/src/themis/message.c.o                           [OK]
compile build/obj/src/themis/secure_cell.c.o                       [OK]
compile build/obj/src/themis/secure_comparator.c.o                 [OK]
compile build/obj/src/themis/secure_keygen.c.o                     [OK]
compile build/obj/src/themis/secure_message.c.o                    [OK]
compile build/obj/src/themis/secure_message_wrapper.c.o            [OK]
compile build/obj/src/themis/secure_session.c.o                    [OK]
compile build/obj/src/themis/secure_session_message.c.o            [OK]
compile build/obj/src/themis/secure_session_peer.c.o               [OK]
compile build/obj/src/themis/secure_session_serialize.c.o          [OK]
compile build/obj/src/themis/secure_session_utils.c.o              [OK]
compile build/obj/src/themis/sym_enc_message.c.o                   [OK]
link build/libthemis.a                                             [OK]
compile build/obj/src/soter/soter_container.c.o                    [OK]
compile build/obj/src/soter/soter_crc32.c.o                        [OK]
compile build/obj/src/soter/soter_ec_key.c.o                       [OK]
compile build/obj/src/soter/soter_hmac.c.o                         [OK]
compile build/obj/src/soter/soter_kdf.c.o                          [OK]
compile build/obj/src/soter/soter_rsa_key.c.o                      [OK]
compile build/obj/src/soter/soter_sign.c.o                         [OK]
compile build/obj/src/soter/ed25519/fe_0.c.o                       [OK]
compile build/obj/src/soter/ed25519/fe_1.c.o                       [OK]
compile build/obj/src/soter/ed25519/fe_add.c.o                     [OK]
compile build/obj/src/soter/ed25519/fe_cmov.c.o                    [OK]
compile build/obj/src/soter/ed25519/fe_copy.c.o                    [OK]
compile build/obj/src/soter/ed25519/fe_frombytes.c.o               [OK]
compile build/obj/src/soter/ed25519/fe_invert.c.o                  [OK]
compile build/obj/src/soter/ed25519/fe_isnegative.c.o              [OK]
compile build/obj/src/soter/ed25519/fe_isnonzero.c.o               [OK]
compile build/obj/src/soter/ed25519/fe_mul.c.o                     [OK]
compile build/obj/src/soter/ed25519/fe_neg.c.o                     [OK]
compile build/obj/src/soter/ed25519/fe_pow22523.c.o                [OK]
compile build/obj/src/soter/ed25519/fe_sq.c.o                      [OK]
compile build/obj/src/soter/ed25519/fe_sq2.c.o                     [OK]
compile build/obj/src/soter/ed25519/fe_sub.c.o                     [OK]
compile build/obj/src/soter/ed25519/fe_tobytes.c.o                 [OK]
compile build/obj/src/soter/ed25519/ge_add.c.o                     [OK]
compile build/obj/src/soter/ed25519/ge_cmp.c.o                     [OK]
compile build/obj/src/soter/ed25519/ge_double_scalarmult.c.o       [OK]
compile build/obj/src/soter/ed25519/ge_frombytes.c.o               [OK]
compile build/obj/src/soter/ed25519/ge_frombytes_no_negate.c.o     [OK]
compile build/obj/src/soter/ed25519/ge_madd.c.o                    [OK]
compile build/obj/src/soter/ed25519/ge_msub.c.o                    [OK]
compile build/obj/src/soter/ed25519/ge_p1p1_to_p2.c.o              [OK]
compile build/obj/src/soter/ed25519/ge_p1p1_to_p3.c.o              [OK]
compile build/obj/src/soter/ed25519/ge_p2_0.c.o                    [OK]
compile build/obj/src/soter/ed25519/ge_p2_dbl.c.o                  [OK]
compile build/obj/src/soter/ed25519/ge_p2_to_p3.c.o                [OK]
compile build/obj/src/soter/ed25519/ge_p3_0.c.o                    [OK]
compile build/obj/src/soter/ed25519/ge_p3_dbl.c.o                  [OK]
compile build/obj/src/soter/ed25519/ge_p3_sub.c.o                  [OK]
compile build/obj/src/soter/ed25519/ge_p3_to_cached.c.o            [OK]
compile build/obj/src/soter/ed25519/ge_p3_to_p2.c.o                [OK]
compile build/obj/src/soter/ed25519/ge_p3_tobytes.c.o              [OK]
compile build/obj/src/soter/ed25519/ge_precomp_0.c.o               [OK]
compile build/obj/src/soter/ed25519/ge_scalarmult.c.o              [OK]
compile build/obj/src/soter/ed25519/ge_scalarmult_base.c.o         [OK]
compile build/obj/src/soter/ed25519/ge_sub.c.o                     [OK]
compile build/obj/src/soter/ed25519/ge_tobytes.c.o                 [OK]
compile build/obj/src/soter/ed25519/gen_rand_32.c.o                [OK]
compile build/obj/src/soter/ed25519/sc_muladd.c.o                  [OK]
compile build/obj/src/soter/ed25519/sc_reduce.c.o                  [OK]
compile build/obj/src/soter/boringssl/soter_asym_cipher.c.o        [OK]
compile build/obj/src/soter/boringssl/soter_asym_ka.c.o            [OK]
compile build/obj/src/soter/boringssl/soter_ec_key.c.o             [OK]
compile build/obj/src/soter/boringssl/soter_ecdsa_common.c.o       [OK]
compile build/obj/src/soter/boringssl/soter_hash.c.o               [OK]
compile build/obj/src/soter/boringssl/soter_rand.c.o               [OK]
compile build/obj/src/soter/boringssl/soter_rsa_common.c.o         [OK]
compile build/obj/src/soter/boringssl/soter_rsa_key.c.o            [OK]
compile build/obj/src/soter/boringssl/soter_rsa_key_pair_gen.c.o   [OK]
compile build/obj/src/soter/boringssl/soter_sign_ecdsa.c.o         [OK]
compile build/obj/src/soter/boringssl/soter_sign_rsa.c.o           [OK]
compile build/obj/src/soter/boringssl/soter_sym.c.o                [OK]
compile build/obj/src/soter/boringssl/soter_verify_ecdsa.c.o       [OK]
compile build/obj/src/soter/boringssl/soter_verify_rsa.c.o         [OK]
compile build/obj/src/soter/boringssl/soter_wipe.c.o               [OK]
link build/libsoter.a                                              [OK]
building embedded BoringSSL...
-- Found Perl: /usr/bin/perl (found version "5.18.4") 
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/frank/git/themis/build/boringssl
Scanning dependencies of target global_target
[  0%] Built target global_target
Scanning dependencies of target fipsmodule
[  0%] Building C object crypto/fipsmodule/CMakeFiles/fipsmodule.dir/bcm.c.o
[  1%] Building C object crypto/fipsmodule/CMakeFiles/fipsmodule.dir/fips_shared_support.c.o
[  1%] Building C object crypto/fipsmodule/CMakeFiles/fipsmodule.dir/is_fips.c.o
[  1%] Built target fipsmodule
[  1%] Generating err_data.c
Scanning dependencies of target crypto
[  3%] Building C object crypto/CMakeFiles/crypto.dir/asn1/a_bitstr.c.o
[  3%] Building C object crypto/CMakeFiles/crypto.dir/asn1/a_bool.c.o
[  3%] Building C object crypto/CMakeFiles/crypto.dir/asn1/a_d2i_fp.c.o
[  3%] Building C object crypto/CMakeFiles/crypto.dir/asn1/a_dup.c.o
[  5%] Building C object crypto/CMakeFiles/crypto.dir/asn1/a_enum.c.o
[  5%] Building C object crypto/CMakeFiles/crypto.dir/asn1/a_gentm.c.o
[  5%] Building C object crypto/CMakeFiles/crypto.dir/asn1/a_i2d_fp.c.o
[  5%] Building C object crypto/CMakeFiles/crypto.dir/asn1/a_int.c.o
[  6%] Building C object crypto/CMakeFiles/crypto.dir/asn1/a_mbstr.c.o
[  6%] Building C object crypto/CMakeFiles/crypto.dir/asn1/a_object.c.o
[  6%] Building C object crypto/CMakeFiles/crypto.dir/asn1/a_octet.c.o
[  6%] Building C object crypto/CMakeFiles/crypto.dir/asn1/a_print.c.o
[  8%] Building C object crypto/CMakeFiles/crypto.dir/asn1/a_strnid.c.o
[  8%] Building C object crypto/CMakeFiles/crypto.dir/asn1/a_time.c.o
[  8%] Building C object crypto/CMakeFiles/crypto.dir/asn1/a_type.c.o
[  8%] Building C object crypto/CMakeFiles/crypto.dir/asn1/a_utctm.c.o
[ 10%] Building C object crypto/CMakeFiles/crypto.dir/asn1/a_utf8.c.o
[ 10%] Building C object crypto/CMakeFiles/crypto.dir/asn1/asn1_lib.c.o
[ 10%] Building C object crypto/CMakeFiles/crypto.dir/asn1/asn1_par.c.o
[ 10%] Building C object crypto/CMakeFiles/crypto.dir/asn1/asn_pack.c.o
[ 11%] Building C object crypto/CMakeFiles/crypto.dir/asn1/f_enum.c.o
[ 11%] Building C object crypto/CMakeFiles/crypto.dir/asn1/f_int.c.o
[ 11%] Building C object crypto/CMakeFiles/crypto.dir/asn1/f_string.c.o
[ 11%] Building C object crypto/CMakeFiles/crypto.dir/asn1/tasn_dec.c.o
[ 13%] Building C object crypto/CMakeFiles/crypto.dir/asn1/tasn_enc.c.o
[ 13%] Building C object crypto/CMakeFiles/crypto.dir/asn1/tasn_fre.c.o
[ 13%] Building C object crypto/CMakeFiles/crypto.dir/asn1/tasn_new.c.o
[ 13%] Building C object crypto/CMakeFiles/crypto.dir/asn1/tasn_typ.c.o
[ 15%] Building C object crypto/CMakeFiles/crypto.dir/asn1/tasn_utl.c.o
[ 15%] Building C object crypto/CMakeFiles/crypto.dir/asn1/time_support.c.o
[ 15%] Building C object crypto/CMakeFiles/crypto.dir/base64/base64.c.o
[ 15%] Building C object crypto/CMakeFiles/crypto.dir/bio/bio.c.o
[ 16%] Building C object crypto/CMakeFiles/crypto.dir/bio/bio_mem.c.o
[ 16%] Building C object crypto/CMakeFiles/crypto.dir/bio/connect.c.o
[ 16%] Building C object crypto/CMakeFiles/crypto.dir/bio/fd.c.o
[ 16%] Building C object crypto/CMakeFiles/crypto.dir/bio/file.c.o
[ 18%] Building C object crypto/CMakeFiles/crypto.dir/bio/hexdump.c.o
[ 18%] Building C object crypto/CMakeFiles/crypto.dir/bio/pair.c.o
[ 18%] Building C object crypto/CMakeFiles/crypto.dir/bio/printf.c.o
[ 18%] Building C object crypto/CMakeFiles/crypto.dir/bio/socket.c.o
[ 20%] Building C object crypto/CMakeFiles/crypto.dir/bio/socket_helper.c.o
[ 20%] Building C object crypto/CMakeFiles/crypto.dir/bn_extra/bn_asn1.c.o
[ 20%] Building C object crypto/CMakeFiles/crypto.dir/bn_extra/convert.c.o
[ 20%] Building C object crypto/CMakeFiles/crypto.dir/buf/buf.c.o
[ 22%] Building C object crypto/CMakeFiles/crypto.dir/bytestring/asn1_compat.c.o
[ 22%] Building C object crypto/CMakeFiles/crypto.dir/bytestring/ber.c.o
[ 22%] Building C object crypto/CMakeFiles/crypto.dir/bytestring/cbb.c.o
[ 22%] Building C object crypto/CMakeFiles/crypto.dir/bytestring/cbs.c.o
[ 23%] Building C object crypto/CMakeFiles/crypto.dir/bytestring/unicode.c.o
[ 23%] Building C object crypto/CMakeFiles/crypto.dir/chacha/chacha.c.o
[ 23%] Building C object crypto/CMakeFiles/crypto.dir/cipher_extra/cipher_extra.c.o
[ 23%] Building C object crypto/CMakeFiles/crypto.dir/cipher_extra/derive_key.c.o
[ 25%] Building C object crypto/CMakeFiles/crypto.dir/cipher_extra/e_aesccm.c.o
[ 25%] Building C object crypto/CMakeFiles/crypto.dir/cipher_extra/e_aesctrhmac.c.o
[ 25%] Building C object crypto/CMakeFiles/crypto.dir/cipher_extra/e_aesgcmsiv.c.o
[ 25%] Building C object crypto/CMakeFiles/crypto.dir/cipher_extra/e_chacha20poly1305.c.o
[ 27%] Building C object crypto/CMakeFiles/crypto.dir/cipher_extra/e_null.c.o
[ 27%] Building C object crypto/CMakeFiles/crypto.dir/cipher_extra/e_rc2.c.o
[ 27%] Building C object crypto/CMakeFiles/crypto.dir/cipher_extra/e_rc4.c.o
[ 27%] Building C object crypto/CMakeFiles/crypto.dir/cipher_extra/e_tls.c.o
[ 28%] Building C object crypto/CMakeFiles/crypto.dir/cipher_extra/tls_cbc.c.o
[ 28%] Building C object crypto/CMakeFiles/crypto.dir/cmac/cmac.c.o
[ 28%] Building C object crypto/CMakeFiles/crypto.dir/conf/conf.c.o
[ 28%] Building C object crypto/CMakeFiles/crypto.dir/cpu-aarch64-fuchsia.c.o
[ 30%] Building C object crypto/CMakeFiles/crypto.dir/cpu-aarch64-linux.c.o
[ 30%] Building C object crypto/CMakeFiles/crypto.dir/cpu-arm-linux.c.o
[ 30%] Building C object crypto/CMakeFiles/crypto.dir/cpu-arm.c.o
[ 30%] Building C object crypto/CMakeFiles/crypto.dir/cpu-intel.c.o
[ 32%] Building C object crypto/CMakeFiles/crypto.dir/cpu-ppc64le.c.o
[ 32%] Building C object crypto/CMakeFiles/crypto.dir/crypto.c.o
[ 32%] Building C object crypto/CMakeFiles/crypto.dir/curve25519/spake25519.c.o
[ 32%] Building C object crypto/CMakeFiles/crypto.dir/dh/dh.c.o
[ 33%] Building C object crypto/CMakeFiles/crypto.dir/dh/params.c.o
[ 33%] Building C object crypto/CMakeFiles/crypto.dir/dh/check.c.o
[ 33%] Building C object crypto/CMakeFiles/crypto.dir/dh/dh_asn1.c.o
[ 33%] Building C object crypto/CMakeFiles/crypto.dir/digest_extra/digest_extra.c.o
[ 35%] Building C object crypto/CMakeFiles/crypto.dir/dsa/dsa.c.o
[ 35%] Building C object crypto/CMakeFiles/crypto.dir/dsa/dsa_asn1.c.o
[ 35%] Building C object crypto/CMakeFiles/crypto.dir/ecdh_extra/ecdh_extra.c.o
[ 35%] Building C object crypto/CMakeFiles/crypto.dir/ecdsa_extra/ecdsa_asn1.c.o
[ 37%] Building C object crypto/CMakeFiles/crypto.dir/ec_extra/ec_asn1.c.o
[ 37%] Building C object crypto/CMakeFiles/crypto.dir/err/err.c.o
[ 37%] Building C object crypto/CMakeFiles/crypto.dir/err_data.c.o
[ 37%] Building C object crypto/CMakeFiles/crypto.dir/engine/engine.c.o
[ 38%] Building C object crypto/CMakeFiles/crypto.dir/evp/digestsign.c.o
[ 38%] Building C object crypto/CMakeFiles/crypto.dir/evp/evp.c.o
[ 38%] Building C object crypto/CMakeFiles/crypto.dir/evp/evp_asn1.c.o
[ 38%] Building C object crypto/CMakeFiles/crypto.dir/evp/evp_ctx.c.o
[ 40%] Building C object crypto/CMakeFiles/crypto.dir/evp/p_dsa_asn1.c.o
[ 40%] Building C object crypto/CMakeFiles/crypto.dir/evp/p_ec.c.o
[ 40%] Building C object crypto/CMakeFiles/crypto.dir/evp/p_ec_asn1.c.o
[ 40%] Building C object crypto/CMakeFiles/crypto.dir/evp/p_ed25519.c.o
[ 42%] Building C object crypto/CMakeFiles/crypto.dir/evp/p_ed25519_asn1.c.o
[ 42%] Building C object crypto/CMakeFiles/crypto.dir/evp/p_rsa.c.o
[ 42%] Building C object crypto/CMakeFiles/crypto.dir/evp/p_rsa_asn1.c.o
[ 42%] Building C object crypto/CMakeFiles/crypto.dir/evp/p_x25519.c.o
[ 44%] Building C object crypto/CMakeFiles/crypto.dir/evp/p_x25519_asn1.c.o
[ 44%] Building C object crypto/CMakeFiles/crypto.dir/evp/pbkdf.c.o
[ 44%] Building C object crypto/CMakeFiles/crypto.dir/evp/print.c.o
[ 44%] Building C object crypto/CMakeFiles/crypto.dir/evp/scrypt.c.o
[ 45%] Building C object crypto/CMakeFiles/crypto.dir/evp/sign.c.o
[ 45%] Building C object crypto/CMakeFiles/crypto.dir/ex_data.c.o
[ 45%] Building C object crypto/CMakeFiles/crypto.dir/hkdf/hkdf.c.o
[ 45%] Building C object crypto/CMakeFiles/crypto.dir/hrss/hrss.c.o
[ 47%] Building C object crypto/CMakeFiles/crypto.dir/lhash/lhash.c.o
[ 47%] Building C object crypto/CMakeFiles/crypto.dir/mem.c.o
[ 47%] Building C object crypto/CMakeFiles/crypto.dir/obj/obj.c.o
[ 47%] Building C object crypto/CMakeFiles/crypto.dir/obj/obj_xref.c.o
[ 49%] Building C object crypto/CMakeFiles/crypto.dir/pem/pem_all.c.o
[ 49%] Building C object crypto/CMakeFiles/crypto.dir/pem/pem_info.c.o
[ 49%] Building C object crypto/CMakeFiles/crypto.dir/pem/pem_lib.c.o
[ 49%] Building C object crypto/CMakeFiles/crypto.dir/pem/pem_oth.c.o
[ 50%] Building C object crypto/CMakeFiles/crypto.dir/pem/pem_pk8.c.o
[ 50%] Building C object crypto/CMakeFiles/crypto.dir/pem/pem_pkey.c.o
[ 50%] Building C object crypto/CMakeFiles/crypto.dir/pem/pem_x509.c.o
[ 50%] Building C object crypto/CMakeFiles/crypto.dir/pem/pem_xaux.c.o
[ 52%] Building C object crypto/CMakeFiles/crypto.dir/pkcs7/pkcs7.c.o
[ 52%] Building C object crypto/CMakeFiles/crypto.dir/pkcs7/pkcs7_x509.c.o
[ 52%] Building C object crypto/CMakeFiles/crypto.dir/pkcs8/pkcs8.c.o
[ 52%] Building C object crypto/CMakeFiles/crypto.dir/pkcs8/pkcs8_x509.c.o
[ 54%] Building C object crypto/CMakeFiles/crypto.dir/pkcs8/p5_pbev2.c.o
[ 54%] Building C object crypto/CMakeFiles/crypto.dir/poly1305/poly1305.c.o
[ 54%] Building C object crypto/CMakeFiles/crypto.dir/poly1305/poly1305_arm.c.o
[ 54%] Building C object crypto/CMakeFiles/crypto.dir/poly1305/poly1305_vec.c.o
[ 55%] Building C object crypto/CMakeFiles/crypto.dir/pool/pool.c.o
[ 55%] Building C object crypto/CMakeFiles/crypto.dir/rand_extra/deterministic.c.o
[ 55%] Building C object crypto/CMakeFiles/crypto.dir/rand_extra/forkunsafe.c.o
[ 55%] Building C object crypto/CMakeFiles/crypto.dir/rand_extra/fuchsia.c.o
[ 57%] Building C object crypto/CMakeFiles/crypto.dir/rand_extra/rand_extra.c.o
[ 57%] Building C object crypto/CMakeFiles/crypto.dir/rand_extra/windows.c.o
[ 57%] Building C object crypto/CMakeFiles/crypto.dir/rc4/rc4.c.o
[ 57%] Building C object crypto/CMakeFiles/crypto.dir/refcount_c11.c.o
[ 59%] Building C object crypto/CMakeFiles/crypto.dir/refcount_lock.c.o
[ 59%] Building C object crypto/CMakeFiles/crypto.dir/rsa_extra/rsa_asn1.c.o
[ 59%] Building C object crypto/CMakeFiles/crypto.dir/rsa_extra/rsa_print.c.o
[ 59%] Building C object crypto/CMakeFiles/crypto.dir/stack/stack.c.o
[ 61%] Building C object crypto/CMakeFiles/crypto.dir/siphash/siphash.c.o
[ 61%] Building C object crypto/CMakeFiles/crypto.dir/thread.c.o
[ 61%] Building C object crypto/CMakeFiles/crypto.dir/thread_none.c.o
[ 61%] Building C object crypto/CMakeFiles/crypto.dir/thread_pthread.c.o
[ 62%] Building C object crypto/CMakeFiles/crypto.dir/thread_win.c.o
[ 62%] Building C object crypto/CMakeFiles/crypto.dir/x509/a_digest.c.o
[ 62%] Building C object crypto/CMakeFiles/crypto.dir/x509/a_sign.c.o
[ 62%] Building C object crypto/CMakeFiles/crypto.dir/x509/a_strex.c.o
[ 64%] Building C object crypto/CMakeFiles/crypto.dir/x509/a_verify.c.o
[ 64%] Building C object crypto/CMakeFiles/crypto.dir/x509/algorithm.c.o
[ 64%] Building C object crypto/CMakeFiles/crypto.dir/x509/asn1_gen.c.o
[ 64%] Building C object crypto/CMakeFiles/crypto.dir/x509/by_dir.c.o
[ 66%] Building C object crypto/CMakeFiles/crypto.dir/x509/by_file.c.o
[ 66%] Building C object crypto/CMakeFiles/crypto.dir/x509/i2d_pr.c.o
[ 66%] Building C object crypto/CMakeFiles/crypto.dir/x509/rsa_pss.c.o
[ 66%] Building C object crypto/CMakeFiles/crypto.dir/x509/t_crl.c.o
[ 67%] Building C object crypto/CMakeFiles/crypto.dir/x509/t_req.c.o
[ 67%] Building C object crypto/CMakeFiles/crypto.dir/x509/t_x509.c.o
[ 67%] Building C object crypto/CMakeFiles/crypto.dir/x509/t_x509a.c.o
[ 67%] Building C object crypto/CMakeFiles/crypto.dir/x509/x509.c.o
[ 69%] Building C object crypto/CMakeFiles/crypto.dir/x509/x509_att.c.o
[ 69%] Building C object crypto/CMakeFiles/crypto.dir/x509/x509_cmp.c.o
[ 69%] Building C object crypto/CMakeFiles/crypto.dir/x509/x509_d2.c.o
[ 69%] Building C object crypto/CMakeFiles/crypto.dir/x509/x509_def.c.o
[ 71%] Building C object crypto/CMakeFiles/crypto.dir/x509/x509_ext.c.o
[ 71%] Building C object crypto/CMakeFiles/crypto.dir/x509/x509_lu.c.o
[ 71%] Building C object crypto/CMakeFiles/crypto.dir/x509/x509_obj.c.o
[ 71%] Building C object crypto/CMakeFiles/crypto.dir/x509/x509_r2x.c.o
[ 72%] Building C object crypto/CMakeFiles/crypto.dir/x509/x509_req.c.o
[ 72%] Building C object crypto/CMakeFiles/crypto.dir/x509/x509_set.c.o
[ 72%] Building C object crypto/CMakeFiles/crypto.dir/x509/x509_trs.c.o
[ 72%] Building C object crypto/CMakeFiles/crypto.dir/x509/x509_txt.c.o
[ 74%] Building C object crypto/CMakeFiles/crypto.dir/x509/x509_v3.c.o
[ 74%] Building C object crypto/CMakeFiles/crypto.dir/x509/x509_vfy.c.o
[ 74%] Building C object crypto/CMakeFiles/crypto.dir/x509/x509_vpm.c.o
[ 74%] Building C object crypto/CMakeFiles/crypto.dir/x509/x509cset.c.o
[ 76%] Building C object crypto/CMakeFiles/crypto.dir/x509/x509name.c.o
[ 76%] Building C object crypto/CMakeFiles/crypto.dir/x509/x509rset.c.o
[ 76%] Building C object crypto/CMakeFiles/crypto.dir/x509/x509spki.c.o
[ 76%] Building C object crypto/CMakeFiles/crypto.dir/x509/x_algor.c.o
[ 77%] Building C object crypto/CMakeFiles/crypto.dir/x509/x_all.c.o
[ 77%] Building C object crypto/CMakeFiles/crypto.dir/x509/x_attrib.c.o
[ 77%] Building C object crypto/CMakeFiles/crypto.dir/x509/x_crl.c.o
[ 77%] Building C object crypto/CMakeFiles/crypto.dir/x509/x_exten.c.o
[ 79%] Building C object crypto/CMakeFiles/crypto.dir/x509/x_info.c.o
[ 79%] Building C object crypto/CMakeFiles/crypto.dir/x509/x_name.c.o
[ 79%] Building C object crypto/CMakeFiles/crypto.dir/x509/x_pkey.c.o
[ 79%] Building C object crypto/CMakeFiles/crypto.dir/x509/x_pubkey.c.o
[ 81%] Building C object crypto/CMakeFiles/crypto.dir/x509/x_req.c.o
[ 81%] Building C object crypto/CMakeFiles/crypto.dir/x509/x_sig.c.o
[ 81%] Building C object crypto/CMakeFiles/crypto.dir/x509/x_spki.c.o
[ 81%] Building C object crypto/CMakeFiles/crypto.dir/x509/x_val.c.o
[ 83%] Building C object crypto/CMakeFiles/crypto.dir/x509/x_x509.c.o
[ 83%] Building C object crypto/CMakeFiles/crypto.dir/x509/x_x509a.c.o
[ 83%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/pcy_cache.c.o
[ 83%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/pcy_data.c.o
[ 84%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/pcy_lib.c.o
[ 84%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/pcy_map.c.o
[ 84%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/pcy_node.c.o
[ 84%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/pcy_tree.c.o
[ 86%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/v3_akey.c.o
[ 86%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/v3_akeya.c.o
[ 86%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/v3_alt.c.o
[ 86%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/v3_bcons.c.o
[ 88%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/v3_bitst.c.o
[ 88%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/v3_conf.c.o
[ 88%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/v3_cpols.c.o
[ 88%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/v3_crld.c.o
[ 89%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/v3_enum.c.o
[ 89%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/v3_extku.c.o
[ 89%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/v3_genn.c.o
[ 89%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/v3_ia5.c.o
[ 91%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/v3_info.c.o
[ 91%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/v3_int.c.o
[ 91%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/v3_lib.c.o
[ 91%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/v3_ncons.c.o
[ 93%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/v3_ocsp.c.o
[ 93%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/v3_pci.c.o
[ 93%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/v3_pcia.c.o
[ 93%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/v3_pcons.c.o
[ 94%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/v3_pku.c.o
[ 94%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/v3_pmaps.c.o
[ 94%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/v3_prn.c.o
[ 94%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/v3_purp.c.o
[ 96%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/v3_skey.c.o
[ 96%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/v3_sxnet.c.o
[ 96%] Building C object crypto/CMakeFiles/crypto.dir/x509v3/v3_utl.c.o
[ 96%] Building C object crypto/CMakeFiles/crypto.dir/__/third_party/fiat/curve25519.c.o
[ 98%] Building C object crypto/CMakeFiles/crypto.dir/__/third_party/sike/fpx.c.o
[ 98%] Building C object crypto/CMakeFiles/crypto.dir/__/third_party/sike/isogeny.c.o
[ 98%] Building C object crypto/CMakeFiles/crypto.dir/__/third_party/sike/curve_params.c.o
[ 98%] Building C object crypto/CMakeFiles/crypto.dir/__/third_party/sike/sike.c.o
[100%] Building C object crypto/CMakeFiles/crypto.dir/__/third_party/sike/asm/fp_generic.c.o
[100%] Linking C static library libcrypto.a
[100%] Built target crypto
[  0%] Built target global_target
[  1%] Built target fipsmodule
[ 80%] Built target crypto
Scanning dependencies of target ssl
[ 80%] Building CXX object ssl/CMakeFiles/ssl.dir/bio_ssl.cc.o
[ 80%] Building CXX object ssl/CMakeFiles/ssl.dir/d1_both.cc.o
[ 82%] Building CXX object ssl/CMakeFiles/ssl.dir/d1_lib.cc.o
[ 82%] Building CXX object ssl/CMakeFiles/ssl.dir/d1_pkt.cc.o
[ 82%] Building CXX object ssl/CMakeFiles/ssl.dir/d1_srtp.cc.o
[ 82%] Building CXX object ssl/CMakeFiles/ssl.dir/dtls_method.cc.o
[ 83%] Building CXX object ssl/CMakeFiles/ssl.dir/dtls_record.cc.o
[ 83%] Building CXX object ssl/CMakeFiles/ssl.dir/handoff.cc.o
[ 83%] Building CXX object ssl/CMakeFiles/ssl.dir/handshake.cc.o
[ 83%] Building CXX object ssl/CMakeFiles/ssl.dir/handshake_client.cc.o
[ 84%] Building CXX object ssl/CMakeFiles/ssl.dir/handshake_server.cc.o
[ 84%] Building CXX object ssl/CMakeFiles/ssl.dir/s3_both.cc.o
[ 84%] Building CXX object ssl/CMakeFiles/ssl.dir/s3_lib.cc.o
[ 84%] Building CXX object ssl/CMakeFiles/ssl.dir/s3_pkt.cc.o
[ 86%] Building CXX object ssl/CMakeFiles/ssl.dir/ssl_aead_ctx.cc.o
[ 86%] Building CXX object ssl/CMakeFiles/ssl.dir/ssl_asn1.cc.o
[ 86%] Building CXX object ssl/CMakeFiles/ssl.dir/ssl_buffer.cc.o
[ 86%] Building CXX object ssl/CMakeFiles/ssl.dir/ssl_cert.cc.o
[ 87%] Building CXX object ssl/CMakeFiles/ssl.dir/ssl_cipher.cc.o
[ 87%] Building CXX object ssl/CMakeFiles/ssl.dir/ssl_file.cc.o
[ 87%] Building CXX object ssl/CMakeFiles/ssl.dir/ssl_key_share.cc.o
[ 87%] Building CXX object ssl/CMakeFiles/ssl.dir/ssl_lib.cc.o
[ 89%] Building CXX object ssl/CMakeFiles/ssl.dir/ssl_privkey.cc.o
[ 89%] Building CXX object ssl/CMakeFiles/ssl.dir/ssl_session.cc.o
[ 89%] Building CXX object ssl/CMakeFiles/ssl.dir/ssl_stat.cc.o
[ 89%] Building CXX object ssl/CMakeFiles/ssl.dir/ssl_transcript.cc.o
[ 90%] Building CXX object ssl/CMakeFiles/ssl.dir/ssl_versions.cc.o
[ 90%] Building CXX object ssl/CMakeFiles/ssl.dir/ssl_x509.cc.o
[ 90%] Building CXX object ssl/CMakeFiles/ssl.dir/t1_enc.cc.o
[ 90%] Building CXX object ssl/CMakeFiles/ssl.dir/t1_lib.cc.o
[ 91%] Building CXX object ssl/CMakeFiles/ssl.dir/tls_method.cc.o
[ 91%] Building CXX object ssl/CMakeFiles/ssl.dir/tls_record.cc.o
[ 91%] Building CXX object ssl/CMakeFiles/ssl.dir/tls13_both.cc.o
[ 91%] Building CXX object ssl/CMakeFiles/ssl.dir/tls13_client.cc.o
[ 93%] Building CXX object ssl/CMakeFiles/ssl.dir/tls13_enc.cc.o
[ 93%] Building CXX object ssl/CMakeFiles/ssl.dir/tls13_server.cc.o
[ 93%] Linking CXX static library libssl.a
[ 93%] Built target ssl
Scanning dependencies of target decrepit
[ 94%] Building C object decrepit/CMakeFiles/decrepit.dir/bio/base64_bio.c.o
[ 94%] Building C object decrepit/CMakeFiles/decrepit.dir/blowfish/blowfish.c.o
[ 94%] Building C object decrepit/CMakeFiles/decrepit.dir/cast/cast.c.o
[ 94%] Building C object decrepit/CMakeFiles/decrepit.dir/cast/cast_tables.c.o
[ 95%] Building C object decrepit/CMakeFiles/decrepit.dir/cfb/cfb.c.o
[ 95%] Building C object decrepit/CMakeFiles/decrepit.dir/des/cfb64ede.c.o
[ 95%] Building C object decrepit/CMakeFiles/decrepit.dir/dh/dh_decrepit.c.o
[ 95%] Building C object decrepit/CMakeFiles/decrepit.dir/dsa/dsa_decrepit.c.o
[ 97%] Building C object decrepit/CMakeFiles/decrepit.dir/evp/dss1.c.o
[ 97%] Building C object decrepit/CMakeFiles/decrepit.dir/evp/evp_do_all.c.o
[ 97%] Building C object decrepit/CMakeFiles/decrepit.dir/obj/obj_decrepit.c.o
[ 97%] Building C object decrepit/CMakeFiles/decrepit.dir/rc4/rc4_decrepit.c.o
[ 98%] Building C object decrepit/CMakeFiles/decrepit.dir/ripemd/ripemd.c.o
[ 98%] Building C object decrepit/CMakeFiles/decrepit.dir/rsa/rsa_decrepit.c.o
[ 98%] Building C object decrepit/CMakeFiles/decrepit.dir/ssl/ssl_decrepit.c.o
[ 98%] Building C object decrepit/CMakeFiles/decrepit.dir/x509/x509_decrepit.c.o
[100%] Building C object decrepit/CMakeFiles/decrepit.dir/xts/xts.c.o
[100%] Linking CXX static library libdecrepit.a
[100%] Built target decrepit
link build/libthemis.js                                            [WARNINGS]
/usr/local/Cellar/emscripten/1.38.44/libexec/emcc -o build/libthemis.js build/libthemis.a build/libsoter.a build/boringssl/crypto/libcrypto.a build/boringssl/decrepit/libdecrepit.a -Lbuild -L/usr/local/lib -Wl,-undefined,error  -s EXTRA_EXPORTED_RUNTIME_METHODS=@/Users/frank/git/themis/src/wrappers/themis/wasm/emscripten/runtime_exports.json -s RESERVED_FUNCTION_POINTERS=1 --pre-js /Users/frank/git/themis/src/wrappers/themis/wasm/emscripten/pre.js
cache:INFO: generating system library: libc.a... (this will be cached in "/Users/frank/.emscripten_cache/wasm-obj/libc.a" for subsequent builds)
cache:INFO:  - ok
cache:INFO: generating system library: libcompiler_rt.a... (this will be cached in "/Users/frank/.emscripten_cache/wasm-obj/libcompiler_rt.a" for subsequent builds)
cache:INFO:  - ok
cache:INFO: generating system library: libc-wasm.a... (this will be cached in "/Users/frank/.emscripten_cache/wasm-obj/libc-wasm.a" for subsequent builds)
cache:INFO:  - ok
cache:INFO: generating system library: libc-extras.a... (this will be cached in "/Users/frank/.emscripten_cache/wasm-obj/libc-extras.a" for subsequent builds)
cache:INFO:  - ok
cache:INFO: generating system library: libdlmalloc.a... (this will be cached in "/Users/frank/.emscripten_cache/wasm-obj/libdlmalloc.a" for subsequent builds)
cache:INFO:  - ok
cache:INFO: generating system library: libpthreads_stub.a... (this will be cached in "/Users/frank/.emscripten_cache/wasm-obj/libpthreads_stub.a" for subsequent builds)
cache:INFO:  - ok
cache:INFO: generating system library: libcompiler_rt_wasm.a... (this will be cached in "/Users/frank/.emscripten_cache/wasm-obj/libcompiler_rt_wasm.a" for subsequent builds)
cache:INFO:  - ok
cache:INFO: generating system library: libc_rt_wasm.a... (this will be cached in "/Users/frank/.emscripten_cache/wasm-obj/libc_rt_wasm.a" for subsequent builds)
cache:INFO:  - ok
cache:INFO: generating system asset: generated_struct_info.json... (this will be cached in "/Users/frank/.emscripten_cache/wasm-obj/generated_struct_info.json" for subsequent builds)
cache:INFO:  - ok
pack build/wasm-themis.tgz npm notice 
npm notice 📦  wasm-themis@0.12.1
npm notice === Tarball Contents === 
npm notice 11.4kB  LICENSE                           
npm notice 1.3kB   src/index.js                      
npm notice 105.5kB src/libthemis.js                  
npm notice 7.8kB   src/secure_cell_context_imprint.js
npm notice 6.7kB   src/secure_cell_seal.js           
npm notice 7.8kB   src/secure_cell_token_protect.js  
npm notice 876B    src/secure_cell.js                
npm notice 6.8kB   src/secure_comparator.js          
npm notice 6.8kB   src/secure_keygen.js              
npm notice 12.9kB  src/secure_message.js             
npm notice 15.8kB  src/secure_session.js             
npm notice 43.4kB  test/test.js                      
npm notice 4.0kB   src/themis_error.js               
npm notice 2.0kB   src/utils.js                      
npm notice 596B    package.json                      
npm notice 1.7kB   README.md                         
npm notice 9.3kB   src/libthemis.wasm                
npm notice === Tarball Details === 
npm notice name:          wasm-themis                             
npm notice version:       0.12.1                                  
npm notice filename:      wasm-themis-0.12.1.tgz                  
npm notice package size:  48.2 kB                                 
npm notice unpacked size: 244.7 kB                                
npm notice shasum:        9b7ca4ec55eddd1a0de362500b2c580d5f65d1c7
npm notice integrity:     sha512-SwoEFIQhSpPhu[...]vaV+fJS3G3Xzg==
npm notice total files:   17                                      
npm notice 
                                                                   [OK]

Reply:
> I build it on macOS Mojave 10.14.6 and emcc version is as follows.
>
> ```
> emcc -v
> emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 1.38.44
> clang version 10.0.0 (/b/s/w/ir/cache/git/chromium.googlesource.com-external-github.com-llvm-llvm--project > b5f295ffcec2fa7402e39eb1262acbd55a7d39f5)
> Target: x86_64-apple-darwin18.7.0
> Thread model: posix
> ```

Oh, poor soul... Now I understand.

Unfortunately, Emscripten’s “upstream” version of toolchain (the one that uses LLVM’s compiler directly) had some bugs and was kinda broken so we have been using the “fastcomp” flavor. Until recently the “fastcomp” version was the one installed by default so we did not expect anyone running into these issues. I’m sorry that you have managed to.

I have reproduced a similar issue with 1.38.44-upstream. The build runs just fine, but no tests pass because

```
  64) wasm-themis
       Secure Session
         handles type mismatches:
     TypeError: libthemis._themis_gen_ec_key_pair is not a function
      at generateECKeyPair (src/secure_keygen.js:128:21)
      at new KeyPair (src/secure_keygen.js:110:27)
      at makeKeyMaterial (test/test.js:689:27)
      at Context.<anonymous> (test/test.js:833:24)
      at processImmediate (internal/timers.js:439:21)



npm ERR! Test failed.  See above for more details.
make: *** [test_wasm] Error 1
```

Could you please try switching to “fastcomp”?

Do this in a new terminal window:

```
$ ~/emsdk/emsdk install 1.38.44-fastcomp
$ ~/emsdk/emsdk activate 1.38.44-fastcomp
$ source ~/emsdk/emsdk_env.sh
```

Then rebuild everything and run the tests again:

```
$ emmake make clean
$ emmake make wasmthemis
$ emmake make test_wasm
```

It should work like that. I’ve checked on my Mojave and it seems to work.

That’s... a shame that this isn’t documented anywhere. I’ll see this information to the docs.

P. S. I’ve also checked recent updates in the 1.39 branch of Emscripten. It looks like “upstream” has been somewhat fixed only recently, since the 1.39.1 version. Before that it is not even able to run the Themis Core test suite compiled with Emscripten.

And even the latest 1.39.4 version fails to run WasmThemis tests (with an error above). It seems that EMSCRIPTEN_KEEPALIVE does not work with LLVM at all. We’ll need to fix this somehow later. For now only “fastcomp” works.

Reply:
> There are two errors at the beginning of build. They are related with "PHP". I have no idea if the issue is the cause.

No, not really. That’s because our Makefile does literally everything for every platform. It probably fails to detect that you do not have PHP installed yet still tries to do some PHP stuff. You can safely ignore these warnings. I’ll see what we can do about them...

Reply:
Yes. This is the root cause. I tried to switch "emsdk" to "fastcomp" and everything works well now. Thank you!